### ISSUE TITLE
[Feature request] Quick contest "mode"

### ISSUE THREAD

**Issue by Grayscalae** [crnormand/gurps#847]
- **Objective**: have the software calculate margins of victory or loss for quick contests

- **How:**
- - Add little "QC" button next to 3d roll button. Make it so its "on" or "off" state is clearly distinguishable. 
- - When turned on, output to chat "A quick contest of..." Or something like that
- - After it's turned on, have it consider the next two immediate rolls as the rolls of the quick contest, whatever actors roll them.
- - After the rolls are made, output to chat the results, say, "Quick contest of X (actor name) vs Y (actor name). 
Actor name is successful. Margin of victory is Z."
- - - Instead of just actor names, perhaps inserting actor portraits as well?
- - Have the margins be clickable, as to add them to the modifier bucket. 

Cheers.

**Comment by crnormand** [comment]
I've thought about this in the past.   It is definitely something I would like to implement (QCs).  Unfortunately, UI space is at a premium, and setting aside space for another button may be hard to do.

However, we might be able to do this as a drag and drop.   The player (or GM), drags a roll (attribute, skill, attack, etc.) onto the token (or character sheet) of their QC opponent.   

For example, for a ST vs ST roll, the GM could drag the ST roll value of the NPC onto the token (or character sheet) of the player character.   This will roll the ST for the NPC quietly, and then send a message to the player.   The player would be notified that they need to make an opposing roll.   Whatever roll the player makes, will be sent back to the GM.   The 2 rolls will be compared, and the margin of success/failure displayed in the chat as a clickable modifier.

We could probably also make it work in the other direction...   if the PC drags a roll onto an NPC token (since they probably cannot see the character sheet), it would roll the skill, and send the results to the GM.

There is an issue of visibility.   The GM might not want the players to see success/failure, or if they don't mind showing that, they might not want them to see the actual margin.

**Comment by Grayscalae** [comment]
I understand the issue of UI space, but if I may suggest, and if at all possible, I'd really reserve the drag and drop roll workflow for targeting and applying damage. Simply drag and drop an weapon skill roll into a token, and it would roll the attack on the token. 

Similarly, drag and drop the damage into a token, and it rolls damage _and_ automatically opens up the ADD for that token. It think it would make things much more agile.

**Comment by crnormand** [comment]
Good points.    

If the GM dragged and dropped a damage roll onto a token/character sheet, it could automatically roll the damage and bring up the ADD.   It might not be as useful for Players, since they don't normally have access to the ADD (The GM must explicitly enable players to use the ADD), but it could be a small time saver for the GM (instead of clicking the damage roll, and then dragging from the chat log to the target).

As for dragging and dropping an attack skill...   if the player drags an attack roll onto a target, it could automatically roll the attack.   And if it succeeds, it would notify the GM that they need to roll an active defense for the target.    Maybe we can offer a dialog that shows:

No Defense, Dodge, Parry, Block, or Quick Contest

Or if the attack failed, it would just pop up:

Do you want to treat this as a Quick Contest?

Unfortunately, this pops up a dialog every time a player makes an attack... hmm... maybe not the best thing.  But if we continued with this idea, the GM could pick any of the defense options, or they could pick Quick Contest, and then roll anything and it would be applied as a Quick Contest versus the Player's roll.

So from the player's standpoint, they don't have to do anything different... they just drag their skill on to a target token.   If the skill happens to be from a melee or ranged attack, the GM would have the option to defend.... or treat as a QC.

Going in the other direction... hmmm.. I don't think I would offer the same functionality.   Mainly because I don't really want to (and I don't think we need to) automate the Player's process.  I want to make the GM's process easier... and if that means a quick dialog to roll dodge, parry or block, then great, because they are probably handling a lot of NPCs.   But the player only (usually) has the one character.... and (I believe) it is part of the game for the Player to figure out their defense, and roll it.    But this is a personal opinion... and could be changed.   

Ok, what I mean to say is that if a GM drags a skill roll to a player's character token/sheet, it would just ask for a Quick Contest.   If the GM is attacking (instead), they would just roll the attack skill as normal, and if it succeeds, ask the player to make a defense roll.    

Ok, I am babbling a bit... which just means the creative juices are flowing.   I think there is something here we can implement to make the game a bit easier for the GM (my primary goal).

**Comment by Grayscalae** [comment]
> If the GM dragged and dropped a damage roll onto a token/character sheet, it could automatically roll the damage and bring up the ADD. 

This would be great.

> It might not be as useful for Players, since they don't normally have access to the ADD (The GM must explicitly enable players to use the ADD), but it could be a small time saver for the GM (instead of clicking the damage roll, and then dragging from the chat log to the target).

I see a couple possibilities here. When a player drags and drops damage roll onto a token:
- Option 1, the default: roll damage _and open up the ADD for the GM_, so he can apply it (this is the least disruptive for the player and the one to be implemented first, IMO).
- Option 2: have the players set the ADD, and then _send_ it to the GM, so he can review it and apply it. This could be done per level of player trust as well.
- Have some level of player (trusted, assistant GM, whatever), be able to interact with the ADD and apply damage to tokens directly.
 
Thinking of it, I think instead of a player trust level decision this should be selected on a per player basis. This would allow for a more fine control.

> As for dragging and dropping an attack skill... if the player drags an attack roll onto a target, it could automatically roll the attack. And if it succeeds, it would notify the GM that they need to roll an active defense for the target. Maybe we can offer a dialog that shows:
> 
> No Defense, Dodge, Parry, Block, or Quick Contest
> 
> Or if the attack failed, it would just pop up:
> 
> Do you want to treat this as a Quick Contest?
> 
> So from the player's standpoint, they don't have to do anything different... they just drag their skill on to a target token. If the skill happens to be from a melee or ranged attack, the GM would have the option to defend.... or treat as a QC.

Having a pop up for selecting defenses would be great too! Would really speed up things for the GM. We could make it so it shows the skill level for each defense right on the pop-up ( Dodge-11, Parry-10 (Shortsword), Block-10 (Medium Shield) - bonus points if the Parry button is actually a drop-down menu with other parry choices - with their scores - on it. 

I agree on the "pops up a dialog every time a player makes an attack" concern. Perhaps, if instead of a pop-up dialog for this workflow we're discussing,, we should move this to the chat window? Like so (text in red just to make it stand out here):

![image](https://user-images.githubusercontent.com/86167546/124183720-0260aa00-da8f-11eb-85a3-eb19fa65cbe7.png)
 
That, it can easily be ignored by the GM. Truth be told, I think this is even better than a pop-up!

> 
> Going in the other direction... hmmm.. I don't think I would offer the same functionality. Mainly because I don't really want to (and I don't think we need to) automate the Player's process. I want to make the GM's process easier... and if that means a quick dialog to roll dodge, parry or block, then great, because they are probably handling a lot of NPCs. But the player only (usually) has the one character.... and (I believe) it is part of the game for the Player to figure out their defense, and roll it. But this is a personal opinion... and could be changed.

I see this as worth discussion and that both approaches have their own merits.

**Comment by mjeffw** [comment]
I don't like the approach of dragging a roll onto a target to do a Quick Contest because that is a limited version of Quick Contest. I think many QCs in GURPS use different stats between attacker and target. Stealth is the most common example -- it's a QC of the stalker's Stealth versus the target's Per/Hearing/Vision. Implementing QC only for the same stat ignores all of the other QCs where the stats are different.

**Comment by Grayscalae** [comment]
> I don't like the approach of dragging a roll onto a target to do a Quick Contest because that is a limited version of Quick Contest. Many QCs in GURPS use different stats between attacker and target. Stealth is the most common example -- it's a QC of the stalker's Stealth versus the target's Per/Hearing/Vision. Implementing QC only for the same stat ignores all of the other QCs where the stats are different.

I don't believe that would necessarily be the case. After selecting Quick Contest, the system could treat the immediately next roll as the second roll of the QC, whatever roll that may be.

**Comment by crnormand** [comment]
My first thought on an implementation would be that if a player or GM drags a roll to a token (or character sheet), then the target player would be presented with a dialog saying that they need to make a contested roll, and display an open box.   The player would then select their opposing roll, and instead of clicking on it, they would drag it into the open box of the dialog.

The system would then roll both values, and present the differences in margins of success.
